mrmustard/physics/wires.py

```diff
@@ -1,31 +1,169 @@
+from dataclasses import dataclass, field, replace
+from random import randint
+from IPython.display import display
+from mrmustard import widgets
 
-#     http://www.apache.org/licenses/LICENSE-2.0
 
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
+__all__ = ["Wires"]
 
-"""``Wires`` class for supporting tensor network functionalities."""
+"""
+This module provides wire functionality for applications in MrMustard.
+It defines the core classes for representing quantum and classical wires, and their
+relationships in quantum optical circuits.
+"""
 
-from __future__ import annotations
-from functools import cached_property
-import numpy as np
 
-from IPython.display import display
+class LegibleEnum(Enum):
+    """Enum class that provides a more legible string representation."""
 
-from mrmustard import widgets
+    def __str__(self) -> str:
+        return self.name
 
-__all__ = ["Wires"]
+    def __repr__(self) -> str:
+        return self.name
+
+
+class ReprEnum(LegibleEnum):
+    """Enumeration of possible representations for quantum states and operations."""
+
+    UNSPECIFIED = auto()
+    BARGMANN = auto()
+    FOCK = auto()
+    QUADRATURE = auto()
+    PHASESPACE = auto()
+    CHARACTERISTIC = auto()
+
+
+class WiresType(LegibleEnum):
+    """Enumeration of possible wire types in quantum circuits."""
+
+    DM_LIKE = auto()  # only output ket and bra on same modes
+    KET_LIKE = auto()  # only output ket
+    UNITARY_LIKE = auto()  # such that can map ket to ket
+    CHANNEL_LIKE = auto()  # such that can map dm to dm
+    PROJ_MEAS_LIKE = auto()  # only input ket
+    POVM_LIKE = auto()  # only input ket and input bra on same modes
+    CLASSICAL_LIKE = auto()  # only classical wires
+
+
+@dataclass(slots=True)
+class QuantumWire:
+    """
+    Represents a quantum wire in a circuit.
+
+    Args:
+        mode: The mode number this wire represents
+        is_out: Whether this is an output wire
+        is_ket: Whether this wire is on the ket side
+        index: The index of this wire in the circuit
+        repr: The representation of this wire
+        id: Unique identifier for this wire
+    """
+
+    mode: int
+    is_out: bool
+    is_ket: bool
+    index: int
+    repr: ReprEnum = ReprEnum.BARGMANN
+    repr_params: Any = None
+    id: int = field(default_factory=lambda: randint(0, 2**32 - 1), compare=False)
+
+    def __hash__(self) -> int:
+        return hash((self.mode, self.is_out, self.is_ket, self.repr))
+
+    def __repr__(self) -> str:
+        return f"QuantumWire(mode={self.mode}, {'out' if self.is_out else 'in'}, {'ket' if self.is_ket else 'bra'}, repr={self.repr}, index={self.index})"
+
+    def __eq__(self, other: QuantumWire) -> bool:
```

Comment:
Dataclasses have built-in equality and hashing generation.  Could this not be achieved by flagging the appropriate fields?

Reply:
good idea, thanks

Reply:
Ah (just looked it up) but we also need to set frozen=True, which we don't want.